### PR TITLE
feat: add boilerplate for asset manager contract

### PR DESCRIPTION
**Pull request by nightowl121**

### CODE REVIEW

**Review by nightowl121 — COMMENTED** [review]
Please check the review comments @BDEVcoder

File: contracts/core-contracts/cw-asset-manager/src/state.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+//equivalent to map(key:(depositor,token) -> amount)
+pub const DEPOSITS: Map<(&Addr, &Addr), Uint128> = Map::new("deposits_of");
+
+//map(user -> vec[token_addr])
+pub const USER_TOKENS: Map<&Addr, Vec<Addr>> = Map::new("user_tokens");
+
+//xcall deployed address on archway
+pub const SOURCE_XCALL: Item<String> = Item::new("source_xcall_address");
+
+pub const ICON_LOANS_ADDRESS: Item<String> = Item::new("destination_btp_address");
```
Comment (nightowl121) [inline contracts/core-contracts/cw-asset-manager/src/state.rs:16]: The key is still not changed.

File: contracts/core-contracts/cw-asset-manager/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+
+    pub fn withdraw_request(
+        deps: DepsMut,
+        info: MessageInfo,
+        token_address: String,
+        amount: Uint128,
+    ) -> Result<Response, ContractError> {
+        //check withdrawer's current token balance
+        let token = deps.api.addr_validate(&token_address)?;
+        if DEPOSITS.has(deps.storage, (&info.sender, &token)) {
```
[the remark this answers is not in the document]
Comment (BDEVcoder) [reply contracts/core-contracts/cw-asset-manager/src/contract.rs]: current deposit balance for token of the user has been checked

File: contracts/core-contracts/cw-asset-manager/src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+
+    //helper function to transfer tokens from contract to account
+    pub fn transfer_tokens(
+        deps: DepsMut,
+        account: String,
+        token_address: String,
+        amount: Uint128,
+    ) -> Result<Response, ContractError> {
+        let account = Addr::unchecked(account);
+        let token_address = Addr::unchecked(token_address);
```
[the remark this answers is not in the document]
Comment (BDEVcoder) [reply contracts/core-contracts/cw-asset-manager/src/contract.rs:288]: i have used address reference for states therefore use unchecked addr for the transfer token fun()

File: contracts/core-contracts/cw-asset-manager/src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,683 @@
...
+        }
+
+        ExecuteMsg::Deposit {
+            token_address,
+            amount,
+            to,
+            data,
+        } => {
+            let nid = NID.load(deps.storage)?;
+            let depositor = NetworkAddress::new(nid.as_str(), info.sender.as_str());
```
Comment (AntonAndell) [inline contracts/core-contracts/cw-asset-manager/src/contract.rs:68]: Does not need to be converted to network address can be kept as regular address

**Review by AntonAndell — CHANGES_REQUESTED** [review]
Some minor changes could be made, otherwise it looks good.
Also RLP lib has to be changed to the one in IBC integrations to keep consistent with JAVA rlp

File: contracts/core-contracts/cw-asset-manager/src/constants.rs [hunk 5]
```diff
@@ -0,0 +1,3 @@
+//Reply messages identifiers
+
+pub const SUCCESS_REPLY_MSG: u64 = 1;
```
[the remark this answers is not in the document]
Comment (nightowl121) [reply contracts/core-contracts/cw-asset-manager/src/constants.rs:3]: The id is supplied as a placeholder, it's not received back.

File: contracts/core-contracts/cw-asset-manager/src/contract.rs (hunk 4 above)
Comment (nightowl121) [reply contracts/core-contracts/cw-asset-manager/src/contract.rs:68]: Agree, will convert it to native archway address.

**Review by AntonAndell — APPROVED** [review]
LGTM, lets get this in and do smaller Upgrade PRs :+1:

**Pull request merged**